### PR TITLE
Load all channels with conda internals

### PR DESCRIPTION
**Pull request by jaimergp**
### Description

Closes https://github.com/conda/conda-libmamba-solver/issues/59

---

We are replacing the libmamba networking stack with conda internals:

- We download all JSONs with the newest `conda.SubdirData.repo_fetch` interface that doesn't process the JSON (only downloads). 
- We use `libmambapy.Repo`, without the `SubdirData` indirection. This interface allow us to use a JSON file directly (without having to be optimistic about the "find the JSON file" logic in `SubdirData`).
- All the tests that checked the libmamba networking stack have been removed, but the utilities are still there in case we ever need them in the future.
- These changes also allow us to drop tons of pre- and post-processing of MatchSpec and PackageRecord objects (e.g. percentage-encoding and decoding on local paths so libcurl doesn't choke, Windows things, etc). The libmamba<->conda index interface is better abstracted now.

---

There's some legacy issues we should remove in the future:

- The `SubdirData` subclass can introduce some issues with incomplete cache entries we have to invalidate (some tests break otherwise), because metaclass-based caches are that fun.
- In some tests, the cached files are not found because the integration tests abuse the `SubdirData._cache_` a bit too much and some of those changes are not reflected on disk. Some changes upstream will alleviate this.
- TODO: 
  - [ ] Change `conda.testing.helpers._patch_for_local_exports` to _not_ patch `_mtime` to `inf`. Return early if the solver is classic because that function is not needed there.
  - [ ] Remove [this block](https://github.com/conda/conda-libmamba-solver/blob/739dd05b0ab1abe523bf8535e3c3f34c556345e6/conda_libmamba_solver/index.py#L229-L234)

### CODE REVIEW

**Comment by jaimergp** [comment]
Remaining errors have to do with `conda.testing.helpers._patch_for_local_exports` again. Since we are using `SubdirData` now, the cached entries get in the way of the testing mocking machinery.

**Comment by jaimergp** [comment]
`tests/test_priority.py::PriorityIntegrationTests::test_channel_order_channel_priority_true` because defaults has Python 3.8.15 but conda-forge does not (yet), so it can't update.

**Comment by jaimergp** [comment]
https://github.com/conda/conda/pull/12104 will also fix some failures.

**Comment by jaimergp** [comment]
Summary of failures:

Unit tests

- [ ] `./conda-libmamba-solver-src/tests/test_solvers.py::TestLibMambaSolver::test_timestamps_and_deps` - unsure, reproducible
- [ ] `./conda-libmamba-solver-src/tests/test_user_agent.py::test_user_agent_conda_info[libmamba]` - fixed by https://github.com/conda/conda/pull/12104
- [ ] `./conda-libmamba-solver-src/tests/test_user_agent.py::test_user_agent_libmamba_packages` - same

Integration tests
- [ ] `./conda/tests/test_priority.py::PriorityIntegrationTests::test_channel_order_channel_priority_true` - should be fixed when conda-forge releases Python 3.8.15, or maybe it's something to do with channel priority configurations.
- [ ] Windows only:
  - [ ] `./conda/tests/test_activate.py::ShellWrapperIntegrationTests::test_cmd_exe_basic_integration` - Plugin DLL cannot be loaded
  - [ ] `./conda/tests/test_activate.py::test_stacking` - maybe the same?
  - [ ] `./conda/tests/core/test_solve.py::test_auto_update_conda` - write permissions issue

**Comment by dholth** [comment]
We've merged a new conda that changes the way repodata.json is dowloaded, and we will change it again in the next release.

**Comment by jaimergp** [comment]
Nope, #12003 changed the internal API so as a matter of facts everything will fail here as it is now. I am inclined to wait til things get calmer around the repodata fetching logic before working on this PR further. What do you think @dholth?

**Comment by jezdez** [comment]
> Nope, [conda/conda#12003](https://github.com/conda/conda/pull/12003) changed the internal API so as a matter of facts everything will fail here as it is now. I am inclined to wait til things get calmer around the repodata fetching logic before working on this PR further. What do you think @dholth?

I concur, let's move this to a later milestone.

**Comment by jezdez** [comment]
@jaimergp Could you try rebasing this PR?

**Comment by jaimergp** [comment]
Done!

**Comment by jaimergp** [comment]
Using a [custom channel](https://anaconda.org/jaimergp/libmamba) to upload builds [coming](https://github.com/jaimergp-feedstocks/mamba-feedstock/tree/debug-repo-add-channel) from [this PR upstream in `mamba-org/mamba`](https://github.com/mamba-org/mamba/pull/2352). That PR exposes the API needed to register the channel via `libmambapy.Repo` so we forget about matching cache implementations, state file and whatnot. It seems to work in my local tests, but I didn't run the whole suite. Let's see 👀

**Comment by jaimergp** [comment]
The errors are either known `libarchive` incompatibilities between conda-forge and defaults, or installation conflicts. Irrelevant here once the upstream PR makes it to a release. The important part is that it fixes the channel unit tests.

There's a local channel test that fails but it looks like an easy solution (URL encoding?):

```
FAILED tests/test_create.py::IntegrationTests::test_install_tarball_from_local_channel
```

To be continued tomorrow.

**Comment by dholth** [comment]
https://github.com/conda/conda/pull/12461 is the new idea for a best repodata interface. You might build SubdirData, get its fetcher, and then ask for the desired repodata format.

**Comment by jaimergp** [comment]
I synced with the mamba team today on their community call and it looks like https://github.com/mamba-org/mamba/pull/2352 will be welcome soon, so we will get unblocked here earlier than expected I'd say!

**Comment by jaimergp** [comment]
Needs https://github.com/conda/conda/pull/12490

**Comment by jaimergp** [comment]
https://github.com/mamba-org/mamba/pull/2398 has been merged! The moment this makes it to a release we can be in a position to merge this.

**Comment by jaimergp** [comment]
```
tests/test_solvers.py::TestLibMambaSolver::test_arch_preferred_over_noarch_when_otherwise_equal 
```

This one is failing sporadically. It can be reproduced locally by rerunning it several times; it will pass 4 out of 5 times or so. I wonder if this has to do with some change in 1.4.1 (this is the first time I see it fail, if memory serves).

File: .github/workflows/upstream_tests.yml [hunk 1]
```diff
@@ -379,7 +379,7 @@ jobs:
     env:
       OS: macOS
       PYTHON: ${{ matrix.python-version }}
-      TEST_SPLITS: ${{ matrix.test-splits }}
+      TEST_SPLITS: 3
```
Comment (dholth) [inline .github/workflows/upstream_tests.yml:382]: `${{ matrix.test-splits}}` is necessary to have 2 unit tests jobs, and 3 integration tests jobs.

Comment (jaimergp) [reply .github/workflows/upstream_tests.yml:382]: Oops, this must have leaked in the multiple rebases.

**Comment by jaimergp** [comment]
I added some stdout in the spinner (so it hides when needed; e.g. --json or --quiet), and it looks like this:

During fetch:

```
$ CONDA_SUBDIR=win-64 conda create -n unused --dry-run python --solver=libmamba -c ../conda-src/tests/data/legacy_repodata/ -c defaults
Channels:
 - file:///opt/conda-src/tests/data/legacy_repodata
 - defaults
 - conda-forge
Platform: win-64
Collect all metadata (repodata.json): /
```

Once done:

```
$ CONDA_SUBDIR=win-64 conda create -n unused --dry-run python --solver=libmamba -c ../conda-src/tests/data/legacy_repodata/ -c defaults
Channels:
 - file:///opt/conda-src/tests/data/legacy_repodata
 - defaults
 - conda-forge
Platform: win-64
Collect all metadata (repodata.json): done
Solving environment: done
```

**Comment by jezdez** [comment]
> 

Nice!

File: tests/test_user_agent.py [hunk 2]
```diff
@@ -8,14 +8,7 @@
...
 
-As of June 2022, we only use libmamba/libcurl to fetch the repodata.json files.
-Package downloads still take place with the usual conda/requests stack. In other
-words:
-
-* If we are using conda classic, we will use requests to download both
-  repodatas and tarballs.
-* If we switch to libmamba solver, we will use curl to download repodatas,
-  but tarballs will still be downloaded with requests.
+From conda-libmamba-solver 22.11, we only use conda internals (requests).
```
Comment (jaimergp) [inline tests/test_user_agent.py:11]:
```suggestion
From conda-libmamba-solver 23.3, we only use conda internals (requests).
```

File: conda_libmamba_solver/solver.py [hunk 3]
```diff
@@ -349,6 +356,11 @@ def _spec_to_str(spec):
             return spec.original_spec_str
         if spec.get("build") and not spec.get("version"):
             spec = MatchSpec(spec, version="*")
+        # if spec.get("channel"):
+        #     parts = dict(spec._match_components)
+        #     channel = parts.pop("channel")._raw_value
+        #     spec_str = str(MatchSpec(**parts))
+        #     return f"{channel.subdir_url}::{spec_str}"
```
Comment (jaimergp) [inline conda_libmamba_solver/solver.py:363]:
```suggestion
```

File: conda_libmamba_solver/index.py [hunk 4]
```diff
@@ -1,22 +1,91 @@
...
+Finally, a channel can be mounted in a local directory and referred to via
+a regular path, or a file:// URL, with or without normalization on Windows.
+
+The approach
+------------
+We pass the subdir-specific, authenticated URLs to patched conda's 'SubdirData' instances,
+which download the JSON files but do not process them to PackageRecords.
+Once the cache has been populated, we can instantiate 'libmamba.Repo' objects directly.
+We maintain a map of subdir-specific URLs to `conda.model.channel.Channel`
+and `libmamba.Repo` objects.
```
Comment (jaimergp) [inline conda_libmamba_solver/index.py:70]:
```suggestion
We pass the subdir-specific, authenticated URLs to conda's 'SubdirData.repo_patch',
which download the JSON files but do not process them to PackageRecords.
Once the cache has been populated, we can instantiate 'libmamba.Repo' objects directly.
We maintain a map of subdir-specific URLs to `conda.model.channel.Channel`
and `libmamba.Repo` objects.
```

**Review by jezdez — APPROVED** [review]

**Pull request merged**